std/path.d

```diff
@@ -691,9 +691,21 @@ if (isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementT
         A slice of $(D path).
 */
 auto rootName(R)(R path)
-if ((isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementType!R) ||
-    isNarrowString!R) &&
-    !isConvertibleToString!R)
+if (isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementType!R) && !isSomeString!R)
+{
+    return _rootName(path);
+}
+
+/// ditto
+auto rootName(C)(C[] path)
+if (isSomeChar!C)
```

Comment:
Hmm - looking at all your PRs this uglifies Phobos a bit without any plan to remove the "redundant" methods in the near future. At least at some point there was a consensus that we should try to go into the opposite direction and reduce the number of overloads, which are visible to the user...
Won't the issues of slicing a static array be fixed with DIP1000?
If not, what about using the second overload just for `isStaticArray!R && isSomeChar!(ElementType!R)` and marking it as deprecated?

Reply:
Ah I see. I was looking for something like:

```d
auto rootName(R)(R path)
if (isRandomAccessRange!R && hasSlicing!R && hasLength!R && isSomeChar!(ElementType!R) && !isStaticArray!R) { ... }

deprecated("rootName needs to be called with strings not static array. Use []")
auto rootName(C)(C[] path)
if (isSomeChar!C) { ... }
```

but as you said that doesn't work :/
I'm still not happy about adding this churn to the code & doc without any foreseeable way of getting rid of it. After all, we were trying to simply the overloads in the public documentation to make them more digest-able for the user.
But I assume you as you went with this approach that you don't have an idea how we could deprecate the `isStaticArray` case in the foreseeable future?

Reply:
_Maybe_ it would be possible to deprecate accepting static arrays and other implicitly convertible types if an explicit wrapper template is used rather than directly declaring a function template, but it's not possible to deprecate accepting implicitly convertible types using function templates, because the overloads which safely accept the implicitly convertible types are the exact same ones which accept strings.

My take on it is that as much as it sucks, we're permanently stuck accepting implicitly convertible types for the functions that were originally written to accept strings and/or which now use `isConvertibleToString`. But for each of these, we're talking one extra overload with a really simple template constraint. This whole mess would have been _so_ much better if static arrays didn't implicitly convert to dynamic arrays, but I couldn't convince Walter to change that the last time I tried. His take on it was that DIP 1000 would solve the safety problem, so it wasn't a problem. :|